File: wallet/modules/native-ratls/ios/NativeRaTlsModule.swift
// Copyright (c) Privasys. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-only

import ExpoModulesCore
import NativeRaTlsC

public class NativeRaTlsModule: Module {
    public func definition() -> ModuleDefinition {
        Name("NativeRaTls")

        AsyncFunction("inspect") { (host: String, port: Int, caCertPath: String?) -> String in
            return await withCheckedContinuation { continuation in
                DispatchQueue.global(qos: .userInitiated).async {
                    let result = host.withCString { hostPtr in
                        if let path = caCertPath {
                            return path.withCString { pathPtr in
                                ratls_inspect(hostPtr, UInt16(port), pathPtr)
                            }
                        } else {
                            return ratls_inspect(hostPtr, UInt16(port), nil)
                        }
                    }

                    guard let result else {
                        continuation.resume(returning: "{\"error\":\"FFI returned null\"}")
                        return
                    }

                    let json = String(cString: result)
                    ratls_free_string(result)
                    continuation.resume(returning: json)
                }
            }
        }

        AsyncFunction("verify") { (host: String, port: Int, caCertPath: String?, policyJson: String) -> String in
            return await withCheckedContinuation { continuation in
                DispatchQueue.global(qos: .userInitiated).async {
                    let result = host.withCString { hostPtr in
                        policyJson.withCString { policyPtr in
                            if let path = caCertPath {
                                return path.withCString { pathPtr in
                                    ratls_verify(hostPtr, UInt16(port), pathPtr, policyPtr)
                                }
                            } else {
                                return ratls_verify(hostPtr, UInt16(port), nil, policyPtr)
                            }
                        }
                    }

                    guard let result else {
                        continuation.resume(returning: "{\"error\":\"FFI returned null\"}")
                        return
                    }

                    let json = String(cString: result)
                    ratls_free_string(result)
                    continuation.resume(returning: json)
                }
            }
        }

        AsyncFunction("post") { (host: String, port: Int, path: String, body: String, headersJson: String?, caCertPath: String?) -> String in
            return await withCheckedContinuation { continuation in
                DispatchQueue.global(qos: .userInitiated).async {
                    let result = host.withCString { hostPtr in
                        path.withCString { pathPtr in
                            body.withCString { bodyPtr in
                                // Resolve the optional CA path and headers-JSON to
                                // C pointers (or NULL) before the single FFI call.
                                func call(_ caPtr: UnsafePointer<CChar>?) -> UnsafeMutablePointer<CChar>? {
                                    if let headersJson {
                                        return headersJson.withCString { hPtr in
                                            ratls_post(hostPtr, UInt16(port), caPtr, pathPtr, bodyPtr, hPtr)
                                        }
                                    }
                                    return ratls_post(hostPtr, UInt16(port), caPtr, pathPtr, bodyPtr, nil)
                                }
                                if let caPath = caCertPath {
                                    return caPath.withCString { caPtr in call(caPtr) }
                                } else {
                                    return call(nil)
                                }
                            }
                        }
                    }

                    guard let result else {
                        continuation.resume(returning: "{\"error\":\"FFI returned null\"}")
                        return
                    }

                    let json = String(cString: result)
                    ratls_free_string(result)
                    continuation.resume(returning: json)
                }
            }
        }

        // Method-generic RA-TLS request (GET, POST, PUT, DELETE, ...).
        // An empty body sends no body (correct for GET/DELETE). Mirrors
        // `post`, calling the ratls_request FFI with an explicit method.
        AsyncFunction("request") { (method: String, host: String, port: Int, path: String, body: String, headersJson: String?, caCertPath: String?) -> String in
            return await withCheckedContinuation { continuation in
                DispatchQueue.global(qos: .userInitiated).async {
                    let result = method.withCString { methodPtr in
                        host.withCString { hostPtr in
                            path.withCString { pathPtr in
                                body.withCString { bodyPtr in
                                    func call(_ caPtr: UnsafePointer<CChar>?) -> UnsafeMutablePointer<CChar>? {
                                        if let headersJson {
                                            return headersJson.withCString { hPtr in
                                                ratls_request(methodPtr, hostPtr, UInt16(port), caPtr, pathPtr, bodyPtr, hPtr)
                                            }
                                        }
                                        return ratls_request(methodPtr, hostPtr, UInt16(port), caPtr, pathPtr, bodyPtr, nil)
                                    }
                                    if let caPath = caCertPath {
                                        return caPath.withCString { caPtr in call(caPtr) }
                                    } else {
                                        return call(nil)
                                    }
                                }
                            }
                        }
                    }

                    guard let result else {
                        continuation.resume(returning: "{\"error\":\"FFI returned null\"}")
                        return
                    }

                    let json = String(cString: result)
                    ratls_free_string(result)
                    continuation.resume(returning: json)
                }
            }
        }
    }
}

File: wallet/modules/native-ratls/android/src/main/java/org/privasys/nativeratls/NativeRaTlsModule.kt
// Copyright (c) Privasys. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-only

package org.privasys.nativeratls

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking

class NativeRaTlsModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("NativeRaTls")

        AsyncFunction("inspect") { host: String, port: Int, caCertPath: String? ->
            runBlocking(Dispatchers.IO) {
                NativeRaTlsBridge.nativeInspect(host, port, caCertPath)
            }
        }

        AsyncFunction("verify") { host: String, port: Int, caCertPath: String?, policyJson: String ->
            runBlocking(Dispatchers.IO) {
                NativeRaTlsBridge.nativeVerify(host, port, caCertPath, policyJson)
            }
        }

        // Arg order matches the JS surface (native-ratls/src/index.ts): the TS
        // `post`/`request` pass headers before caCertPath, so the bridge call
        // re-orders them to the FFI (ca, path, body, headers) shape.
        AsyncFunction("post") { host: String, port: Int, path: String, body: String, headersJson: String?, caCertPath: String? ->
            runBlocking(Dispatchers.IO) {
                NativeRaTlsBridge.nativePost(host, port, caCertPath, path, body, headersJson)
            }
        }

        AsyncFunction("request") { method: String, host: String, port: Int, path: String, body: String, headersJson: String?, caCertPath: String? ->
            runBlocking(Dispatchers.IO) {
                NativeRaTlsBridge.nativeRequest(method, host, port, caCertPath, path, body, headersJson)
            }
        }
    }
}
